### PR TITLE
OHRM5X-2088: Refactor Admin FE unique validation to BE validation

### PR DESCRIPTION
**Pull request by devishke-orange**
PR contains:
- OHRM5X-2088
- OHRM5X-2099

Rebased https://github.com/orangehrm/orangehrm/pull/1628

### CODE REVIEW

File: src/plugins/orangehrmAdminPlugin/Dao/ValidationUniqueDao.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+     * @param string $entityName
+     * @param string $attributeName
+     * @param string|null $entityId
+     * @param string|null $matchByField
+     * @param string|null $matchByValue
+     * @return bool
+     */
+    public function isValueUnique(string $value, string $entityName, string $attributeName, ?string $entityId, ?string $matchByField, ?string $matchByValue): bool
+    {
+        $qb = $this->createQueryBuilder('OrangeHRM\\Entity\\'  . ucfirst($entityName), 'entity');
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Dao/ValidationUniqueDao.php:37]:
```suggestion
        $qb = $this->createQueryBuilder('OrangeHRM\\Entity\\'  . $entityName, 'entity');
```
if this is not a big deal, will use exact entity name in the FE, what do u think? :thinking:

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: src/plugins/orangehrmAdminPlugin/Dao/ValidationUniqueDao.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            $qb->andWhere($qb->expr()->neq('entity.' . $id, ':id'))
+                ->setParameter('id', $entityId);
+        }
+
+        if (!is_null($matchByValue) && !is_null($matchByField)) {
+            $qb->andWhere($qb->expr()->eq('entity.' . $matchByField, ':matchValue'))
+                ->setParameter('matchValue', $matchByValue);
+        }
+
+        $results = $qb->getQuery()->execute();
```
[the remark this answers is not in the document]
Comment (devishke-orange) [reply src/plugins/orangehrmAdminPlugin/Dao/ValidationUniqueDao.php:52]: yeah this is simpler. will implement.

File: src/plugins/orangehrmAdminPlugin/Dao/ValidationUniqueDao.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+     * @param string $attributeName
+     * @param string|null $entityId
+     * @param string|null $matchByField
+     * @param string|null $matchByValue
+     * @return bool
+     */
+    public function isValueUnique(string $value, string $entityName, string $attributeName, ?string $entityId, ?string $matchByField, ?string $matchByValue): bool
+    {
+        $qb = $this->createQueryBuilder('OrangeHRM\\Entity\\'  . ucfirst($entityName), 'entity');
+        $qb->andWhere($qb->expr()->eq('entity.' . $attributeName, ':value'))
```
[the remark this answers is not in the document]
Comment (devishke-orange) [reply src/plugins/orangehrmAdminPlugin/Dao/ValidationUniqueDao.php:38]: did not check for SQL injection. will implement the whitelisting suggestion from your other comment

File: src/plugins/orangehrmAdminPlugin/Dao/ValidationUniqueDao.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+            ->setParameter('value', $value);
+
+        if (!is_null($entityId)) {
+            $id = $entityName == 'employee' ? 'empNumber' : 'id';
```
[the remark this answers is not in the document]
Comment (devishke-orange) [reply src/plugins/orangehrmAdminPlugin/Dao/ValidationUniqueDao.php:42]: only one that I came across so far. all others have ID as the primary key. will double check

File: src/plugins/orangehrmAdminPlugin/config/routes.yaml [hunk 5]
```diff
@@ -334,6 +334,14 @@ apiv2_admin_i18n_groups:
   defaults:
     _api: OrangeHRM\Admin\Api\I18NGroupAPI
 
+apiv2_admin_validate_uniqueness:
+  path: /api/v2/admin/validation/unique
```
[the remark this answers is not in the document]
Comment (devishke-orange) [reply src/plugins/orangehrmAdminPlugin/config/routes.yaml:338]: yes did  not define data group. will do so

File: src/plugins/orangehrmAdminPlugin/test/Api/ValidationUniqueAPITest.php [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+use OrangeHRM\Admin\Api\ValidationUniqueAPI;
+use OrangeHRM\Framework\Services;
+use OrangeHRM\Tests\Util\EndpointIntegrationTestCase;
+use OrangeHRM\Tests\Util\Integration\TestCaseParams;
+
+/**
+ * @group Admin
+ * @group APIv2
+ */
+class ValidationUniqueAPITest extends EndpointIntegrationTestCase
```
[the remark this answers is not in the document]
Comment (devishke-orange) [reply src/plugins/orangehrmAdminPlugin/test/Api/ValidationUniqueAPITest.php:31]: will add

File: src/plugins/orangehrmAdminPlugin/Api/ValidationUniqueAPI.php [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+    public function getValidationRuleForGetOne(): ParamRuleCollection
+    {
+        return new ParamRuleCollection(
+            new ParamRule(
+                self::PARAMETER_VALUE,
+                new Rule(Rules::LENGTH, [null, self::PARAM_RULE_VALUE_MAX_LENGTH])
+            ),
+            new ParamRule(
+                self::PARAMETER_ENTITY_NAME,
+                new Rule(Rules::LENGTH, [null, self::PARAM_RULE_ENTITY_NAME_MAX_LENGTH])
```
[the remark this answers is not in the document]
Comment (devishke-orange) [reply src/plugins/orangehrmAdminPlugin/Api/ValidationUniqueAPI.php:142]: will implement this

File: src/plugins/orangehrmAdminPlugin/Dao/ValidationUniqueDao.php (hunk 1 above)
Comment (devishke-orange) [reply src/plugins/orangehrmAdminPlugin/Dao/ValidationUniqueDao.php:37]: sure. I'll change it

File: src/plugins/orangehrmAdminPlugin/Dao/ValidationUniqueDao.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (devishke-orange) [reply src/plugins/orangehrmAdminPlugin/Dao/ValidationUniqueDao.php]: other entities also have. only one we are validating for now is `Employee`. I added a new argument to provide the ID field if necessary

File: src/client/src/orangehrmAdminPlugin/pages/systemUser/EditSystemUser.vue [hunk 8]
```diff
@@ -225,29 +233,6 @@ export default {
           this.onCancel();
         });
     },
-    validateUserName(user) {
```
Comment (RajithaKumara) [inline src/client/src/orangehrmAdminPlugin/pages/systemUser/EditSystemUser.vue:228]: I feel better to revert this and reuse `/api/v2/admin/validation/user-name`, and will remove this entity from the new API

File: src/client/src/orangehrmAdminPlugin/pages/systemUser/SaveSystemUser.vue [hunk 9]
```diff
@@ -178,28 +185,6 @@ export default {
           this.onCancel();
         });
     },
-    validateUserName(user) {
```
Comment (RajithaKumara) [inline src/client/src/orangehrmAdminPlugin/pages/systemUser/SaveSystemUser.vue:181]: I feel better to revert this and reuse `/api/v2/admin/validation/user-name`, and will remove this entity from the new API

File: src/client/src/orangehrmPimPlugin/pages/employee/SaveEmployee.vue [hunk 10]
```diff
@@ -205,13 +220,17 @@ export default {
...
+        ],
         empPicture: [
           maxFileSize(1024 * 1024),
           validFileTypes(this.allowedImageTypes),
         ],
         username: [
           required,
+          this.usernameUniqueValidation,
           shouldNotLessThanCharLength(5),
           shouldNotExceedCharLength(40),
```
Comment (RajithaKumara) [inline src/client/src/orangehrmPimPlugin/pages/employee/SaveEmployee.vue:235]:
```suggestion
          required,
          shouldNotLessThanCharLength(5),
          shouldNotExceedCharLength(40),
          this.usernameUniqueValidation,
```
:thinking:  I feel this should be the order. first check client side validation first and then only we should pass to the BE. what do you think

File: src/client/src/orangehrmTimePlugin/components/AddCustomerModal.vue [hunk 11]
```diff
@@ -106,28 +117,6 @@ export default {
     onCancel() {
       this.$emit('close');
     },
-    validateCustomerName(customer) {
-      return new Promise((resolve) => {
-        if (customer) {
-          this.http
-            .request({
-              method: 'GET',
-              url: `/api/v2/time/validation/customer-name`,
```
Comment (RajithaKumara) [inline src/client/src/orangehrmTimePlugin/components/AddCustomerModal.vue:115]: Oops. seems here also we have a API

File: src/client/src/orangehrmTimePlugin/pages/customer/EditCustomer.vue [hunk 12]
```diff
@@ -135,29 +146,6 @@ export default {
     onCancel() {
       navigate('/time/viewCustomers');
     },
-    validateCustomerName(customer) {
-      return new Promise((resolve) => {
-        if (customer) {
-          this.http
-            .request({
-              method: 'GET',
-              url: `/api/v2/time/validation/customer-name`,
```
Comment (RajithaKumara) [inline src/client/src/orangehrmTimePlugin/pages/customer/EditCustomer.vue:144]: *

File: src/client/src/orangehrmTimePlugin/pages/project/ProjectEdit.vue [hunk 13]
```diff
@@ -267,28 +271,16 @@ export default {
         });
     },
     validateProjectName(project) {
-      return new Promise((resolve) => {
-        if (project) {
-          this.http
-            .request({
-              method: 'GET',
-              url: `/api/v2/time/validation/project-name`,
```
Comment (RajithaKumara) [inline src/client/src/orangehrmTimePlugin/pages/project/ProjectEdit.vue:275]: Oops. seems here also we have a API

File: src/client/src/orangehrmTimePlugin/pages/project/ProjectSave.vue [hunk 14]
```diff
@@ -214,27 +217,15 @@ export default {
         });
     },
     validateProjectName(project) {
-      return new Promise((resolve) => {
-        if (project) {
-          this.http
-            .request({
-              method: 'GET',
-              url: `/api/v2/time/validation/project-name`,
```
Comment (RajithaKumara) [inline src/client/src/orangehrmTimePlugin/pages/project/ProjectSave.vue:222]: *

File: src/plugins/orangehrmCorePlugin/Dao/ValidationUniqueDao.php [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+     */
+    public function isValueUnique(string $value, string $entityName, string $attributeName, ?string $entityId, ?string $entityIdField, ?string $matchByField, ?string $matchByValue): bool
+    {
+        $qb = $this->createQueryBuilder('OrangeHRM\\Entity\\'  . $entityName, 'entity');
+        $qb->andWhere($qb->expr()->eq('entity.' . $attributeName, ':value'))
+            ->setParameter('value', $value);
+
+        if (!is_null($entityId)) {
+            $id = $entityIdField ?? 'id';
+            $qb->andWhere($qb->expr()->neq('entity.' . $id, ':id'))
```
Comment (RajithaKumara) [inline src/plugins/orangehrmCorePlugin/Dao/ValidationUniqueDao.php:42]: *

File: src/plugins/orangehrmCorePlugin/Dao/ValidationUniqueDao.php [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+                ->setParameter('id', $entityId);
+        }
+
+        if (!is_null($matchByValue) && !is_null($matchByField)) {
+            $qb->andWhere($qb->expr()->eq('entity.' . $matchByField, ':matchValue'))
```
Comment (RajithaKumara) [inline src/plugins/orangehrmCorePlugin/Dao/ValidationUniqueDao.php:47]: I'm not sure this and above string concatenation to `entity.` will introduce SQL injection as `$matchByField` can be a user input according to my understanding.
you can further check https://cheatsheetseries.owasp.org/cheatsheets/SQL_Injection_Prevention_Cheat_Sheet.html

File: installer/Migration/V5_6_0/permission/api.yaml [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+  api: OrangeHRM\Core\Api\Rest\ValidationUniqueAPI
+  module: core
+  allowed:
+    read: true
+    create: false
+    update: false
+    delete: false
+  permissions:
+    - { role: Admin, permission: { read: true, create: false, update: false, delete: false } }
+    - { role: ESS, permission: { read: true, create: false, update: false, delete: false } }
```
Comment (RajithaKumara) [inline installer/Migration/V5_6_0/permission/api.yaml:16]:
```suggestion
    - { role: ESS, permission: { read: false, create: false, update: false, delete: false } }
```
:thinking: seems all places related to Admin screens. Allowing ESS users to check uniqueness will introduce enumeration attack possibility https://www.upguard.com/blog/what-is-an-enumeration-attack

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: src/client/src/orangehrmPimPlugin/pages/employee/SaveEmployee.vue (hunk 10 above)
Comment (devishke-orange) [reply src/client/src/orangehrmPimPlugin/pages/employee/SaveEmployee.vue:235]: I had it like that initially but due to the debounce the length validation would not work properly for certain cases

File: src/plugins/orangehrmAdminPlugin/Dao/ValidationUniqueDao.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (devishke-orange) [reply src/plugins/orangehrmAdminPlugin/Dao/ValidationUniqueDao.php]: whitelisting added for `entityName`, `attributeName`, `matchByField`. For latter two they are based off the Entity (i.e. only fields valid for that entity is allowed)

File: src/client/src/orangehrmAdminPlugin/pages/systemUser/EditSystemUser.vue (hunk 8 above)
Comment (devishke-orange) [reply src/client/src/orangehrmAdminPlugin/pages/systemUser/EditSystemUser.vue]: reverted

File: src/client/src/orangehrmAdminPlugin/pages/systemUser/SaveSystemUser.vue (hunk 9 above)
Comment (devishke-orange) [reply src/client/src/orangehrmAdminPlugin/pages/systemUser/SaveSystemUser.vue]: reverted

File: src/client/src/orangehrmTimePlugin/components/AddCustomerModal.vue (hunk 11 above)
Comment (devishke-orange) [reply src/client/src/orangehrmTimePlugin/components/AddCustomerModal.vue]: reverted

File: src/client/src/orangehrmTimePlugin/pages/customer/EditCustomer.vue (hunk 12 above)
Comment (devishke-orange) [reply src/client/src/orangehrmTimePlugin/pages/customer/EditCustomer.vue]: reverted

File: src/client/src/orangehrmTimePlugin/pages/project/ProjectEdit.vue (hunk 13 above)
Comment (devishke-orange) [reply src/client/src/orangehrmTimePlugin/pages/project/ProjectEdit.vue]: reverted

File: src/client/src/orangehrmTimePlugin/pages/project/ProjectSave.vue (hunk 14 above)
Comment (devishke-orange) [reply src/client/src/orangehrmTimePlugin/pages/project/ProjectSave.vue]: reverted

File: src/plugins/orangehrmCorePlugin/Dao/ValidationUniqueDao.php (hunk 15 above)
Comment (devishke-orange) [reply src/plugins/orangehrmCorePlugin/Dao/ValidationUniqueDao.php]: `entityIdFIeld` removed

File: src/plugins/orangehrmCorePlugin/Dao/ValidationUniqueDao.php (hunk 16 above)
Comment (devishke-orange) [reply src/plugins/orangehrmCorePlugin/Dao/ValidationUniqueDao.php:45]: `matchByField` is now validated in API

File: installer/Migration/V5_6_0/permission/api.yaml (hunk 17 above)
Comment (devishke-orange) [reply installer/Migration/V5_6_0/permission/api.yaml]: removed API from ESS accessible screens and changed permission

File: src/lib/config/Config.php [hunk 18]
```diff
@@ -51,7 +51,7 @@ class Config
     public const MODE_DEMO = 'demo';
 
     public const PRODUCT_NAME = 'OrangeHRM OS';
-    public const PRODUCT_VERSION = '5.5';
+    public const PRODUCT_VERSION = '5.6';
     public const ORANGEHRM_API_VERSION = '2.5.0';
```
Comment (RajithaKumara) [inline src/lib/config/Config.php:55]:
```suggestion
    public const ORANGEHRM_API_VERSION = '2.6.0';
```

**Review by RajithaKumara — APPROVED** [review]

File: src/lib/config/Config.php (hunk 18 above)
Comment (devishke-orange) [reply src/lib/config/Config.php]: changed

**Pull request merged**